File: aeongseo/BOJ/CPP/boj4949.cpp
/*** 4949. 균형잡힌 세상 ***/

#include<iostream>
#include<string>
#include<stack>
using namespace std;

int main() {
	string txt;
	while (getline(cin, txt)) {	// 입력이 들어오면 반복
		if (txt == ".") break;	// 종료조건인 . 하나만 들어오면 반복 종료

		string ans = "yes";	// 답 초기값 yes로 설정

		stack<int> s;
		for (int i = 0; i < txt.length(); i++) {
			if (txt[i] == '(' || txt[i] == '[') {	// 여는 괄호면 push
				s.push(txt[i]);
			}
			else if (txt[i] == ')' || txt[i] == ']') {
				if (s.size() == 0) {	// 닫는 괄호와 매치할 괄호가 없다면 no 저장, 종료
					ans = "no";
					break;
				}
				else {
					if (txt[i] == ')' && s.top() == '(') {	// 소괄호가 짝이 맞는다면 pop
						s.pop();
						continue;
					}
					else if (txt[i] == ']' && s.top() == '[') { // 대괄호가 짝이 맞는다면 pop
						s.pop();
						continue;
					}
					else {	// 짝이 맞지 않는다면 no 저장, 종료
						ans = "no";
						break;
					}
				}
			}
		}
		if (s.size() != 0) {	// 문자열이 끝났는데 스택에 괄호가 남아있다면 no 저장
			ans = "no";
		}

		cout << ans << '\n';
	}
	return 0;
}

File: yujeong/BOJ/boj4949.py
# 4949. 균형잡힌 세상
from collections import deque

left = ['[', '(']
right = [']', ')']

while True:
    stack = deque()
    string = input()

    # 입력받은 문자열이 '.'이 아니면 진행
    if string != ".":
        ans = "yes"
        for s in string:
            if s in left:   # 왼쪽 괄호면 stack에 넣기 
                stack.append(s)
            elif s in right:    # 오른쪽 괄호면
                if not stack:   # 스택이 비어있으면 균형 x
                    ans = "no"
                    break
                lp = stack.pop()
                if right.index(s) != left.index(lp):    # 짝이 맞지 않는 괄호면 균형 x
                    ans = "no"
                    break
        else:   # 문자열 끝까지 다 봤는데 스택에 괄호가 남아있으면 균형 x
            if stack:
                ans = "no"
        
        print(ans)

    # 입력받은 문자열이 '.'이면 종료 
    else:
        break